Display Block Information by matching block variations

ntsekouras opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
This PR wants to solve the problem of displaying the proper information of a block, if that block had been created from a `block variation`.

This is achieved with:
1. A new hook (`useMatchingVariationInformation`) which receives a `clientId` and  returns the appropriate block information to display.
2. An addition of a property (`isActive`) in Block Variation's settings. This optional property is a function to match block variations. It doesn't check dynamically all block properties to try to find a match, because in many cases it doesn't make sense. An example is `embed` block where we can have changed the `responsive` attribute, so a match would not be found.




The changes that have been made affect the `Block Card (Inspector Tools)`,  `Navigation List View (top bar)` and `Breadcrumbs`.



To make use of the new hook and handle `Transforms icon`(in toolbar), I need to make `BlockSwitcher` a function component which is handled in this PR: https://github.com/WordPress/gutenberg/pull/27674 to make reviewing easier.
The icon in BlockSwitcher will be handled in a follow up after the refactoring PR lands.


![pocVariationsDisplayMatcher](https://user-images.githubusercontent.com/16275880/100781013-27b30a00-3413-11eb-94d8-2b1d360470d6.gif)



## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ntsekouras committed:
getBlockTypeWithVariationInfo selector, BlockDescription component and first changes to display the proper info

ntsekouras commented:
This is a comment of @talldan in another draft PR of mine for this problem and I'm just putting it here to have the discussions.

>@ntsekouras This reminds me of a discussion some of us had about the legacy widget block. That block is similar to embeds in that it has variations for the different types of widgets represented. A subset of widgets are 'reference widgets', which can only be used once in a 'post'. The challenge was making the block supports multiple setting work for variations, which was pretty much impossible without being able to reliably identify the variation used to create the block:

This is the issue - #25494

ntsekouras commented:
@talldan I think if this approach goes on, you could make use of `variationMatcher`.

>With an addition of a property (variationMatcher) in Blocks settings. This optional property is a function to match block variations. It doesn't check dynamically all block properties to try to find a match, because in many cases like it doesn't make sense. An example is embed block where we can have changed the responsive attribute, so a match would not be found.

mcsf commented on packages/block-editor/src/components/block-card/index.js:
```diff
@@ -2,15 +2,20 @@
+	const props = ( clientId && { clientId } ) || blockType;
 	return (
 		<div className="block-editor-block-card">
-			<BlockIcon icon={ icon } showColors />
+			<BlockIcon { ...props } showColors />
 			<div className="block-editor-block-card__content">
-				<h2 className="block-editor-block-card__title">{ title }</h2>
+				<h2 className="block-editor-block-card__title">
+					<BlockTitle { ...props } />
+				</h2>
 				<span className="block-editor-block-card__description">
-					{ description }
+					<BlockDescription { ...props } />
```
I don't like that we're making `BlockIcon` and `BlockDescription` so polymorphic. It quickly becomes difficult to understand which level in the element tree is responsible for providing which attributes. At the level of the project, we shouldn't burden developers with choosing between passing `clientId` or explicit values.

gziolo replied:
The initial idea I had was to unify the way how those leaf components are used. In particular, I wanted to make it easier to use in a different context where you could apply some more advanced modifications inside the component:

```js
return (
	<div className="block-editor-block-card">
		<BlockIcon clientId={ clientId } showColors />
		<div className="block-editor-block-card__content">
			<BlockTitle clientId={ clientId } as="h2" className="block-editor-block-card__title">
			<BlockDescription clientId={ clientId } className="block-editor-block-card__description" />
		</div>
	</div>
);
```

In practice, the same `BlockCard` component is used not only in the block sidebar but also in the block preview in the inserter where you don't have access to `clientId` yet.

I personally feel like developers shouldn't be concerned about deciding how for a given block to get the icon (applies to the title and description as well) and pass it to the matching component. It should accept either `clientId` or `name` and do the rest. If you think that the opposite makes more sense also because of the backward compatibility then we will have to find a way to ensure that correct data is being passed in all places. The biggest challenge is the ambiguity we introduce by having `getBlockType` that returns the original block definition and the new method that combines that with the active block variation that can be only achieved by providing the `clientId`.

ntsekouras replied:
>At the level of the project, we shouldn't burden developers with choosing between passing clientId or explicit values.

I personally don't think it's will be a burden to devs, quite the opposite.. This way we have a consistent way(using the three components) of displaying block's info and could even be more easy to use with the fallback of getting the `selectedBlockClientId`. For example if we need to show info on a selected block it could be just `<BlockDescription />`.

mcsf replied:
I think you two make good arguments, leaving me undecided. I have a follow-up question, but let me know if it's too farfetched:

What if `BlockIcon`, BlockTitle`, `BlockDescription` are made to be contextual:

```jsx
<BlockEditContextProvider value={ ... }> /* implicit ancestor */
    <BlockIcon showColors />
    <BlockTitle as="h2" />
```

mcsf replied:
(edit: I accidentally sent this comment before finishing it)

I think you two make good arguments, leaving me undecided. I have a follow-up question, but let me know if it's too farfetched:

What if `BlockIcon`, `BlockTitle`, `BlockDescription` are made to be contextual:

```jsx
<BlockEditContextProvider value={ ... }> /* implicit ancestor */

        <div className="block-editor-block-card">
            <BlockIcon showColors />
            <BlockTitle as="h2" />
```

... with the provision that `BlockIcon` accepts an explicit `blockName` prop as an override:

```js
BlockIcon = ( { blockName } ) => {
    const { clientId } = useBlockEditContext();
    const icon = blockName ?
        grabIconFromBlockType( blockName ) :
        grabIconFromBlockInstance( clientId );
    /* ... */
```

That way, `BlockCard` might choose to provide its own context in order to address:

> In practice, the same BlockCard component is used not only in the block sidebar but also in the block preview in the inserter where you don't have access to clientId yet.

```js
BlockCard = ( props ) => {
    const element = (
        <div className="block-editor-block-card">
            <BlockIcon showColors />
            <BlockTitle as="h2" />
        </div>
    );

    if ( empty( props ) ) {
        return element;
    }

    return (
        <BlockEditContextProvider value={ props }>
             { element }
        </BlockEditContextProvider>       
    );
}
```

gziolo replied:
I wouldn't mind if we would take advantage of `useBlockEditContext` here, we do it in other places in `@wordpress/block-editor` since it holds both the block's `name` and `clientId`.

https://github.com/WordPress/gutenberg/blob/18349c49bab0d9a227e2608a66420f7a9f6e70ac/packages/block-editor/src/components/block-edit/context.js#L12-L16

ntsekouras replied:
Me neither! I can try these changes, along with some of the other comments and take it from there.. :)

ntsekouras replied:
Me neither! I can try these changes, along with some of the other comments and take it from there.. :)

mcsf reviewed (commented):
I like that we're exploring this, but it's a delicate thing to get right. I've left some comments, but I still haven't decided which approach I prefer (e.g. new dedicated selectors, refactor base `getBlockType` selectors, let certain UI components manage it, etc.).

I also haven't commented on the extension of the Variations API with `variationMatcher`, but I'd like us to weigh it against different options (e.g. `blockType.variationAttributeName`, `blockType.variations[0].isActive()`, etc.).

ntsekouras commented on packages/block-editor/src/components/index.js:
```diff
@@ -91,6 +91,7 @@ export {
 export { default as BlockSettingsMenu } from './block-settings-menu';
 export { default as BlockSettingsMenuControls } from './block-settings-menu-controls';
 export { default as BlockTitle } from './block-title';
+export { default as BlockDescription } from './block-description';
```
I think if we follow this or with some changes direction we should expose it. It should look like @gziolo example:
`<BlockTitle clientId={ clientId } as="h2" className="block-ed....." />` to handle the Title, Description and Icon in a consistent way.

ntsekouras commented on packages/blocks/src/store/selectors.js:
```diff
@@ -95,6 +106,44 @@ export function getBlockVariations( state, blockName, scope ) {
+		// Label will often fall back to the title if no label is defined for the
+		// current label context. We do not want "Paragraph: Paragraph".
+		if ( label !== blockType.title ) {
+			blockType.title = `${ blockType.title }: ${ truncate( label, {
+				length: 15,
+			} ) }`;
+		}
```
You're absolutely right about that.

ntsekouras commented on packages/blocks/src/store/selectors.js:
```diff
@@ -95,6 +106,44 @@ export function getBlockVariations( state, blockName, scope ) {
 }
 
+// TODO jsdoc and experimental and tests
+export const getBlockTypeWithVariationInfo = createRegistrySelector(
```
I think so too, but mostly because of the `label`. It probably should be better to handle label where needed, but other than that I don't believe using two stores are a lot..

Before combining them I had a selector in `block` store similar to `getBlockType` but also checking for variations and each of the components (BlockIcon etc..) would retrieve the other info (attributes) from the `block-editor` store. The exact same code..

ntsekouras commented:
Thanks for starting to look at it and sharing your feedback @gziolo and @mcsf - really appreciate it. This is a problem that we would need to solve (matching variation) and by discussing it, will find the best approach!

I know that the current status of the PR (as of writing this comment) is not the best, but it's a start :) I'd like some feedback from others as well that have worked in related things. --cc @talldan, @youknowriad , @noahtallen

ntsekouras committed:
revert BlockCard component and make the passed attributes top level with depracation

ntsekouras committed:
remove BlockDescription

ntsekouras committed:
change InserterPreviewPanel to use BlockCard with passing top level properties

ntsekouras committed:
create new custom hook for block's display information

ntsekouras committed:
use new hook in BlockInspector

ntsekouras committed:
revert BlockIcon and augment BlockTitle

ntsekouras committed:
Navigation List View update

ntsekouras committed:
remove previous selector

ntsekouras committed:
refactor BlockSwitcher to function component and use the new hook

ntsekouras committed:
change hook

ntsekouras committed:
use CreateSelector and declare WPBlockDisplayInformation typedef

gziolo commented:
> I also haven't commented on the extension of the Variations API with `variationMatcher`, but I'd like us to weigh it against different options (e.g. `blockType.variationAttributeName`, `blockType.variations[0].isActive()`, etc.).

I personally lean towards `blockType.variations[0].isActive()`. This API is specific to variations so it might be difficult to explain in documentation when added as a top-level block type property. While it might create some code duplication when added to individual variations but it's perfectly fine to use a shared function. Besides, it gives the opportunity to select which variations should be considered for the check and even it can be customized per individual item.

I was also wondering, whether the `getBlockDisplayInformation` selector still necessary now that there is a hook that consolidates the same functionality? @youknowriad, can you share some recommendations in terms of performance as it needs to use data from two stores: `core/blocks` and `core/block-editor`. It would be also recomputed every time a block attribute changes, a block type or its block variations.

ntsekouras commented:
I have created a separate PR to refactor `BlockSwitcher` there, to make this one easier to review: https://github.com/WordPress/gutenberg/pull/27674.

>I personally lean towards blockType.variations[0].isActive(). This API is specific to variations so it might be difficult to explain in documentation when added as a top-level block type property. While it might create some code duplication when added to individual variations but it's perfectly fine to use a shared function. Besides, it gives the opportunity to select which variations should be considered for the check and even it can be customized per individual item.

I believe that the main use case of matching block variations doesn't need much customization per variation. I can't think of an example that could utilize that, but of course there could be...

What seems to me important though is:
>This API is specific to variations so it might be difficult to explain in documentation when added as a top-level block type property.

So I'm going to make it per variation.

ntsekouras committed:
revert BlockSwitcher to handle in separate PR

ntsekouras committed:
remove getBlockDisplayInformation selector

ntsekouras commented:
>I was also wondering, whether the getBlockDisplayInformation selector still necessary now that there is a hook that consolidates the same functionality?

I removed the selector. I'll need to revisit `BlockTitle` though because we have some logic there that needs the `blockType` and `attributes` for `getBlockLabel`. I haven't decided how to handle this yet...

ntsekouras committed:
add `isActive` API for block variations to use for trying to find a match

ntsekouras committed:
doc for `isActive` in block variation

ntsekouras committed:
jsdoc fix

gziolo commented on packages/block-editor/src/components/block-title/index.js:
```diff
@@ -44,22 +49,16 @@ export default function BlockTitle( { clientId } ) {
+	// If label is defined we prioritize it over possible possible
+	// block variation match title.
+	if ( label !== blockType.title ) {
+		return `${ blockType.title }: ${ truncate( label, { length: 15 } ) }`;
```
Should it use the block information here as well?

```suggestion
		return `${ blockInformation.title }: ${ truncate( label, { length: 15 } ) }`;
```

ntsekouras replied:
I'm not really sure.. I want to revisit all this `label` functionality, but I want to do it in a follow up to investigate properly and not include possibly even more changes in this PR.

For now by leaving it as is `blockType.title` it keeps the exact previous functionality for cases that `label` was used.

gziolo replied:
Sure, I just wanted to flag it.

gziolo commented on packages/block-editor/src/components/use-matching-variation-information/index.js:
```diff
@@ -0,0 +1,59 @@
+		},
+		[ clientId ]
+	);
+	const blockType = getBlockType( name );
```
Both `getBlockType` and `getBlockVariations` are only tiny wrappers over `core/blocks` datastore calls targeted mostly for plugin authors. It would be more convenient to use direct store references inside `useSelect` calls. It should allow running all the logic inside `useSelect` call.

gziolo commented on packages/blocks/src/api/registration.js:
```diff
@@ -126,6 +133,16 @@ import { store as blocksStore } from '../store';
+ * @typedef {Object} WPBlockDisplayInformation
+ *
+ * @property {string}      title       Human-readable block type label.
+ * @property {JSX.Element} icon        Block type icon.
```
Block types acceppt `WPIcon` that is quite complex structure:
```suggestion
 * @property {WPIcon|JSX.Element} icon        Block type icon.
```

gziolo commented on packages/blocks/src/api/registration.js:
```diff
@@ -126,6 +133,16 @@ import { store as blocksStore } from '../store';
  */
 
+/**
+ * A subset of `WPBlock` type. Contains basic block's information for display reasons.
```
It feels more like part of `@wordpress/block-editor` package. Are you sure it should be promoted as a subset of `WPBlock` type? Isn't it a type that exists only for UI purposes that is completely different?

ntsekouras replied:
You're probably right. I'm thinking of naming it differently, move it to `block-editor` and have it return the `attributes` and `blockType` as well. That would simplify the usage of `getBlockLabel` at least in `BlockTitle` and maybe in more places when we'll check this `label` functionality`.

What do you think of having it return these extra props (attributes,blockType)?

gziolo replied:
I don't know much about `getBlockLabel`, can we leave it out for now?

ntsekouras replied:
Yeah, sure!

gziolo commented on packages/blocks/src/api/registration.js:
```diff
@@ -97,6 +97,13 @@ import { store as blocksStore } from '../store';
  * @property {string[]} [keywords]             An array of terms (which can be translated)
  *                                             that help users discover the variation
  *                                             while searching.
+ * @property {Function} [isActive]			   A function that accepts a block's attributes
```
There is some issue with indentation when browsing on GitHub.

gziolo commented on packages/block-library/src/embed/variations.js:
```diff
@@ -339,4 +339,15 @@ const variations = [
+ */
+variations.forEach( ( variation ) => {
+	if ( variation.isActive ) return;
+	variation.isActive = ( { providerNameSlug } ) =>
```
It would be more convenient when `variationAttributes` would be passed as the second param so people don't have to repeat hard coded values when they define this function in the variation object.

```js
isActive( blockAttributes, variationAttributes ) {
     return blockAttributes.providerNameSlug === variationAttributes.providerNameSlug;
},
```

gziolo reviewed (commented):
Nice work, I think it's close 👍🏻 

There are some things to clarify:
- what should be the final API for `isActive` matcher added to the block variation
- how should `useMatchingVariationInformation` be called, the type proposed is `WPBlockDisplayInformation` so it needs some alignment, it feels like `useBlockDisplayInformation` would better express intent
- we should measure the performance implications of `useMatchingVariationInformation`, at the moment it probably will recompute every time `core/block-editor` changes which would be good, we should ensure it doesn't re-render when other stores change.

ntsekouras committed:
address review feedback part 1

ntsekouras committed:
fix BlockTitle tests

ntsekouras committed:
rename hook to `useBlockDisplayInformation`

ntsekouras committed:
fix export

ntsekouras committed:
useBlockDisplayInformation README

ntsekouras commented:
>what should be the final API for isActive matcher added to the block variation

Changed like this per suggestion: https://github.com/WordPress/gutenberg/pull/27469#discussion_r544145212

>how should useMatchingVariationInformation be called, the type proposed is WPBlockDisplayInformation so it needs some alignment, it feels like useBlockDisplayInformation would better express intent

Makes sense and changed it to `useBlockDisplayInformation`

gziolo commented on packages/block-editor/src/components/use-block-display-information/index.js:
```diff
@@ -0,0 +1,71 @@
+
+// TODO write tests
+export default function useBlockDisplayInformation( clientId ) {
+	const { attributes, blockType, variations } = useSelect(
```
Is it possible to move all logic to `useSelect`?

```js
export default function useBlockDisplayInformation( clientId ) {
    return useSelect(
        // all logic here
    );
}
```

ntsekouras replied:
moved here: https://github.com/WordPress/gutenberg/pull/27469/commits/a42d2b22288484aebf42b6dea8ac30656bbab7ed

ntsekouras committed:
move login in useSelect

ntsekouras committed:
e2e tests

ntsekouras committed:
blockTitle extra safeguard that existed before

gziolo commented on packages/block-editor/src/components/index.js:
```diff
@@ -118,6 +118,7 @@ export {
 export { default as Warning } from './warning';
 export { default as WritingFlow } from './writing-flow';
 export { useCanvasClickRedirect as __unstableUseCanvasClickRedirect } from './use-canvas-click-redirect';
+export { default as __experimentalUseBlockDisplayInformation } from './use-block-display-information';
```
Any reason why it should be experimental? I personally can't think of anything missing so it cab be considered final.

gziolo replied:
Now that happened, [CHANGELOG](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/CHANGELOG.md) could benefit from a note as well. It doesn't feel like it's updated frequently :(

gziolo commented on packages/block-editor/src/components/use-block-display-information/README.md:
```diff
@@ -0,0 +1,44 @@
+# useBlockDisplayInformation
+
+A React Hook that tries to find a matching block variation and returns the appropriate information for display reasons. In order to try to find a match we need to things:
```
Typo:
```suggestion
A React Hook that tries to find a matching block variation and returns the appropriate information for display reasons. In order to try to find a match we need two things:
```

gziolo commented on packages/block-editor/src/components/use-block-display-information/README.md:
```diff
@@ -0,0 +1,44 @@
+
+A React Hook that tries to find a matching block variation and returns the appropriate information for display reasons. In order to try to find a match we need to things:
+
+1. Block's client id to extract it's current attributes.
```
Typo:
```suggestion
1. Block's client id to extract its current attributes.
```

gziolo commented on packages/block-editor/src/components/use-block-display-information/README.md:
```diff
@@ -0,0 +1,44 @@
+A React Hook that tries to find a matching block variation and returns the appropriate information for display reasons. In order to try to find a match we need to things:
+
+1. Block's client id to extract it's current attributes.
+2. A block variation should have set `isActive` prop to a proper function.
```
How do you like the following version?
```suggestion
2. A block variation has `isActive` prop defined with a matcher function.
```

gziolo commented on packages/block-editor/src/components/use-block-display-information/README.md:
```diff
@@ -0,0 +1,44 @@
+
+### Usage
+
+The hook returns an object which contains block's title, icon and description. If no blockType is found with the provided `clientId`, returns `null`.
```
A proposed revision to this sentence:
```suggestion
The hook returns an object which contains the block's title, icon, and description. If no block type is found for the provided `clientId`, it returns `null`.
```

gziolo commented on packages/block-editor/src/components/use-block-display-information/README.md:
```diff
@@ -0,0 +1,44 @@
+	const blockInformation = useBlockDisplayInformation( clientId );
+	const { title, icon, description } = blockInformation;
+	return (
+		<div style={ resizeStyles }>
```
Is it some copy/paste issue?
```suggestion
		<div>
```

gziolo commented on packages/block-editor/src/components/use-block-display-information/index.js:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @param {string} clientId Block's client id.
+ * @return {?WPBlockDisplayInformation} Block's display information.
+ *
```
Nit, this line is not needed.
```suggestion
```

gziolo commented on packages/block-editor/src/components/use-block-display-information/index.js:
```diff
@@ -0,0 +1,66 @@
+ * If no blockType is found with the provided clientId, returns null.
+ *
+ * @param {string} clientId Block's client id.
+ * @return {?WPBlockDisplayInformation} Block's display information.
```
Should it be more specific? Example:
```suggestion
 * @return {?WPBlockDisplayInformation} Block's display information, or `null` when the block or its type not found.
```

gziolo commented on packages/block-editor/src/components/use-block-display-information/index.js:
```diff
@@ -0,0 +1,66 @@
+		( select ) => {
+			if ( ! clientId ) return null;
+			const { getBlockName, getBlockAttributes } = select(
+				'core/block-editor'
```
Can we use the store object rather than the hardcoded string? It's planned for the existing code in #27088 anyway.

gziolo commented on packages/block-editor/src/components/use-block-display-information/index.js:
```diff
@@ -0,0 +1,66 @@
+				icon: blockType.icon,
+				description: blockType.description,
+			};
+			if ( ! variations ) return blockTypeInfo;
```
Using store action you can remove variations so this check should account for an empty array as well. Technically, it would still work but there is some further processing that can be avoided.

ntsekouras replied:
This is a good catch! 👍

gziolo reviewed (commented):
This PR is in excellent shape now so I left mostly nitpicks. Thank you for all iterations. I'm personally satisfied with the last revision proposed.  I still need to test this functionality although I anticipate it being only a formal step because e2e tests added cover that 👍🏻 

I will wait for your response before doing the final check and approve this PR.

ntsekouras committed:
Update docs

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

ntsekouras committed:
Update docs

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

ntsekouras committed:
Update docs

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

ntsekouras committed:
Update docs

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

ntsekouras committed:
rewordings

ntsekouras committed:
remove experimental status

ntsekouras committed:
small refactoring

ntsekouras commented:
Thanks for all the help here @gziolo! I addressed all the feedback and it's waiting for the final check to land 🎉  😄

gziolo commented:
There is one issue I discovered, the icon in the block toolbar is coming from the block type rather than from the variation:

<img width="1126" alt="Screen Shot 2020-12-23 at 13 58 58" src="https://user-images.githubusercontent.com/699132/102998556-a89f8600-4527-11eb-8119-9b104fe98235.png">

<img width="1126" alt="Screen Shot 2020-12-23 at 13 58 58" src="https://user-images.githubusercontent.com/699132/102998344-49417600-4527-11eb-90c8-a77099f8f0e2.png">

In all other places, I'm aware of, everything works perfectly fine. 

I did some testing with the profiler I couldn't notice any unnecessary re-renders in the block card when I was changing unrelated block attributes.

Stats from the performance e2e tests look good as well:

<img width="634" alt="Screen Shot 2020-12-23 at 13 55 09" src="https://user-images.githubusercontent.com/699132/102998546-a0dfe180-4527-11eb-8eac-0ad883f6ee10.png">

<img width="574" alt="Screen Shot 2020-12-23 at 13 55 14" src="https://user-images.githubusercontent.com/699132/102998549-a3423b80-4527-11eb-82b2-ec921a4b12cd.png">

ntsekouras commented:
>There is one issue I discovered, the icon and its label in the block toolbar is coming from the block type rather than from the variation:

Thanks for testing performance impact for this! 💯 

@gziolo I have it mind and that's the reason. I will create a follow up as soon as this one lands. Sounds okay with you?


From my PR description:
>To make use of the new hook and handle Transforms icon(in toolbar), I need to make BlockSwitcher a function component which is handled in this PR: #27674 to make reviewing easier.
The icon in BlockSwitcher will be handled in a follow up after the refactoring PR lands.

gziolo reviewed (approved):
🚢
